File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.go
// ((*()) -> true
// (()*) -> true
// (*)) -> true
// (()* -> true
// *(()))*
// **((
func checkValidString(s string) bool {
    balance := 0
    wildCard := 0

    for i:=0;i<len(s);i++{
        if s[i] == '(' {
            balance+=1
        }else if s[i] == ')' {
            if balance > 0 {
                balance -= 1
            }else if wildCard > 0 {
                wildCard -= 1
            }else{
                return false
            }
        }else if s[i] == '*' {
            wildCard += 1
        }
    }
    if balance > 0 && balance > wildCard {
        return false
    }

    balance = 0
    wildCard = 0

    for i:=len(s)-1;i>=0;i--{
        if s[i] == ')' {
            balance+=1
        }else if s[i] == '(' {
            if balance > 0 {
                balance -= 1
            }else if wildCard > 0 {
                wildCard -= 1
            }else{
                return false
            }
        }else if s[i] == '*' {
            wildCard += 1
        }
    }
    if balance > 0 && balance > wildCard {
        return false
    }
    return true
}

File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.rs
impl Solution {
    pub fn check_valid_string(s: String) -> bool {
        return !Solution::isOverOpenBracket(&s) && !Solution::isOverClosedBracket(&s)
    }

    fn isOverClosedBracket(s: &String) -> bool {
        let mut balance = 0;
        let mut wild = 0;

        for b in s.chars() {
            if b == '(' {
                balance += 1;
            }else if b == ')' {
                if balance > 0 {
                    balance -= 1;
                }else if wild > 0 {
                    wild -= 1
                }else{
                    return true
                }
            }else if b == '*' {
                wild += 1;
            }
        }
        return false
    }

    fn isOverOpenBracket(s: &String) -> bool {
        let mut balance = 0;
        let mut wild = 0;

        for b in s.chars().rev() {
            if b == ')' {
                balance += 1;
            }else if b == '(' {
                if balance > 0 {
                    balance -= 1;
                }else if wild > 0 {
                    wild -= 1
                }else{
                    return true
                }
            }else if b == '*' {
                wild += 1;
            }
        }
        return false
    }
}